### PR TITLE
Separate key command controller and add Toolbar component

### PR DESCRIPTION
**Pull request by benbriggs**
### Separate out key command listener logic from `Editor` and turn it into a reusable higher-order component

Also adds a basic `Toolbar` component useful for building Medium-like editor toolbars detached from the `Editor` component. Together with `KeyCommandListener` it can be used to have full control over any key commands coming from the editor. Some notable omissions from things available within the Toolbar context are `focus`, `blur`, `getReadOnly`, and `setReadOnly`. Maybe in the future they could also be moved out of `Editor` but I don't think that makes as much sense as `KeyCommandController`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -110,3 +110,48 @@ const EditorWithPlugins = plugins(Editor);
...
+  return (
+    <div>
+      <WrappedEditor
+        editorState={editorState}
+        onChange={onChange}
+        handleKeyCommand={handleKeyCommand}
+        addKeyCommandListener={addKeyCommandListener}
+        removeKeyCommandListener={removeKeyCommandListener}
+      />
+      <Toolbar
```
Comment (jciruolo) [inline README.md]: You want this to be `WrappedToolbar`, right?

Comment (benbriggs) [inline README.md]: good catch, yup!

File: src/components/Editor.js [hunk 2]
```diff
@@ -213,23 +168,34 @@ export default React.createClass({
...
+      onChange,
+      addKeyCommandListener,
+      removeKeyCommandListener
+    } = this.props;
+
     const decoratedState = this.getDecoratedState();
 
     return this.props.buttons.map((Button, index) => {
       return (
         <Button
```
Comment (cyvuong) [inline src/components/Editor.js]: I think this needs a `key` prop.

File: src/components/KeyCommandController.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  componentWillMount() {
+    this.keyCommandListeners = List(this.props.keyCommandListeners);
+  },
+
+  addKeyCommandListener(listener) {
+    const {addKeyCommandListener} = this.props;
+
+    if (addKeyCommandListener) {
+      addKeyCommandListener(listener);
+      return;
```
Comment (cyvuong) [inline src/components/KeyCommandController.js]: Could just do an else case for this instead of explicit return.

File: src/components/KeyCommandController.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      onChange,
+      keyCommandListeners, // eslint-disable-line no-unused-vars
+      ...others
+    } = this.props;
+
+
+    return (
+      <Component
+        {...others}
+        ref="editor"
```
Comment (cyvuong) [inline src/components/KeyCommandController.js]: I thought string refs were being deprecated in favor of promoting the `ref={c => this._editor = c}` pattern?

**Comment by cyvuong** [comment]
With this new change allowing people to render the Toolbar in any component, I wonder if it would be worth accepting a prop in `Editor` to have it not render the plugin buttons, but have it still render it by default if it is not specified? Would save some effort of having to hide the included one if you know you're going to use the `Toolbar` component yourself (plus, saves having to rerender it all the time for no reason).

File: src/components/KeyCommandController.js (hunk 3 above)
Comment (benbriggs) [inline src/components/KeyCommandController.js]: i think `return`s are usually preferred over an `else`

File: src/components/KeyCommandController.js (hunk 4 above)
Comment (benbriggs) [inline src/components/KeyCommandController.js]: they are, that's probably a change that should be made across the project in a separate PR however

**Comment by benbriggs** [comment]
yeah good call on the option to hide the toolbar. i was thinking a prop to distinguish whether a button is attached to the overlay or not might be useful too

**Comment by benbriggs** [comment]
feeling good about this, so i'm going to merge 🎉
